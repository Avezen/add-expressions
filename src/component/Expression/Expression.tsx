import React from 'react';

export interface ExpressionProps {
    expression: any;
    setExpression: any;
}

export const Expression = ({expression, setExpression}: ExpressionProps) => {

    const addExpressionPart = () => {
        const item = {factor: 0, power: 0};
        let newFirstExpression = [...expression];
        newFirstExpression.push(item);

        setExpression(newFirstExpression);
    };

    const removeExpressionPart = (key: any) => {
        let newFirstExpression = [...expression];
        newFirstExpression.splice(key, 1);

        setExpression(newFirstExpression);
    };

    const changeExpressionFactor = (factor: any, power: any) => (e: any) => {
        const newFirstExpression = expression.map((obj: any) =>
            obj.power === power ? { ...obj, factor: parseInt(e.target.value) } : obj
        );

        setExpression(newFirstExpression);
    };

    const changeExpressionPower = (factor: any, power: any, key: number) => (e: any) => {
        let newFirstExpression = [...expression];

        newFirstExpression[key].power = parseInt(e.target.value);

        setExpression(newFirstExpression);
    };

    return (
        <div>
            <button id={'add-expression'} onClick={addExpressionPart}>
                +
            </button>
            <br/>
            <div className={'expression'}>
            {expression.map(({factor, power}: {factor: any, power: any}, key: any) => {
                return(
                    <div key={key}>
                        <input id={`factor-${key}`} type={'number'} value={factor} onChange={changeExpressionFactor(factor, power)}/>
                        x
                        <input id={`power-${key}`} type={'number'} value={power} onChange={changeExpressionPower(factor, power, key)}/>
                        <button id={`remove-expression-${key}`} onClick={() => removeExpressionPart(key)}>
                            x
                        </button>
                        <br/>
                    </div>
                )
            })}
            </div>
        </div>
    );
};
