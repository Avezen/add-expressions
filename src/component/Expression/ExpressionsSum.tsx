import React from 'react';

export interface ExpressionsSumProps {
    sumOfExpressions: any;
}

export const ExpressionsSum = ({sumOfExpressions}: ExpressionsSumProps) => (
    <div className={'expression'} style={{display: 'flex', justifyContent: 'center'}}>
        {Object.entries(sumOfExpressions).reverse().map(
            ([power, factor]) => {
                if (power === '0') {
                    return (
                        <>
                            <span>{factor as number > 0 && '+'}</span>{factor as string}
                        </>
                    )
                }

                if (power === '1') {
                    return (
                        <>
                            <span>{factor as number > 0 && '+'}</span>{factor}x
                        </>
                    )
                }

                return (
                    <>
                        <span>{factor as number > 0 && '+'}</span>{factor}x
                        <sup>
                            <small>
                                {power}
                            </small>
                        </sup>
                    </>
                )
            }
        )}
    </div>
);