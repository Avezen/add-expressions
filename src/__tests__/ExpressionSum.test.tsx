import React from "react";
import renderer from 'react-test-renderer';
import {ExpressionsSum, ExpressionsSumProps} from "../component/Expression/ExpressionsSum";


const renderExpressionSum = (props: Partial<ExpressionsSumProps> = {}) => {
    const defaultProps: ExpressionsSumProps = {
        sumOfExpressions: {1: 1},
    };

    return <ExpressionsSum {...defaultProps} {...props} />;
};

describe("<ExpressionsSum />", () => {
    it("should render component", () => {
        const findByTestId = renderExpressionSum(),
            ExpressionsSumComponent = renderer.create(findByTestId).toJSON();

        expect(ExpressionsSumComponent).toMatchSnapshot();
    });
});
