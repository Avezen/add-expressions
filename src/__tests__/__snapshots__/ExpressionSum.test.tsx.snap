// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpressionsSum /> should render component 1`] = `
<div
  className="expression"
  style={
    Object {
      "display": "flex",
      "justifyContent": "center",
    }
  }
>
  <span>
    +
  </span>
  1
  x
</div>
`;
